flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java

```diff
@@ -295,22 +303,25 @@ public String toString() {
 				// and set the slot state to releasing so that it gets eventually freed
 				tasks.values().forEach(task -> task.failExternally(cause));
 			}
+
 			final CompletableFuture<Void> cleanupFuture = FutureUtils
 				.waitForAll(tasks.values().stream().map(TaskSlotPayload::getTerminationFuture).collect(Collectors.toList()))
-				.thenRun(() -> {
-					verifyMemoryFreed();
-					this.memoryManager.shutdown();
-				});
-
+				.thenRun(memoryManager::shutdown);
+			verifyAllManagedMemoryIsReleasedAfter(cleanupFuture);
```

Comment:
Maybe add a test which ensures that we can call `MemoryManager.verifyEmpty()` after it has been shut down. That way we explicitly state that this an explicit contract we want to guard.